Proofs/Experiments/SHA512MemoryAliasing.lean

```diff
@@ -67,15 +67,18 @@ theorem sha512_block_armv8_prelude_sym_ctx_access (s0 : ArmState)
   (h_s0_ctx : read_mem_bytes 64 (ctx_addr s0) s0 = SHA2.h0_512.toBitVec)
   (h_s0_ktbl : read_mem_bytes (SHA2.k_512.length * 8) ktbl_addr s0 = BitVec.flatten SHA2.k_512)
   -- (FIXME) Add separateness invariants for the stack's memory region.
+  -- @bollu: can we assume that `h_s1_ctx_input_separate`
+  -- will be given as ((num_blocks s1).toNat * 128)?
+  -- This is much more harmonious since we do not need to worry about overflow.
   (h_s0_ctx_input_separate :
-    mem_separate (ctx_addr s0)   (ctx_addr s0 + 63)
-                 (input_addr s0) (input_addr s0 + (num_blocks s0 * 128)))
+    mem_separate' (ctx_addr s0)   64
+                 (input_addr s0) ((num_blocks s0).toNat * 128))
   (h_s0_ktbl_ctx_separate :
-    mem_separate (ctx_addr s0)   (ctx_addr s0 + 63)
-                  ktbl_addr      (ktbl_addr + (SHA2.k_512.length * 8 - 1)))
+    mem_separate' (ctx_addr s0) 64
+                  ktbl_addr  (SHA2.k_512.length * 8))
   (h_s0_ktbl_input_separate :
-    mem_separate (input_addr s0) (input_addr s0 + (num_blocks s0 * 128))
-                  ktbl_addr      (ktbl_addr + (SHA2.k_512.length * 8 - 1)))
+    mem_separate' (input_addr s0) ((num_blocks s0).toNat * 128)
+                  ktbl_addr      (SHA2.k_512.length * 8))
```

Comment:
Note that we change the definitions to use `mem_separate'` since that's what the tactic expects. A careful review of this is appreciated @alexkeizer @shigoel

Reply:
LGTM!